core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java

```diff
@@ -939,10 +944,21 @@ public static <C extends Comparable<C>> Range<C> rangeFromJson(
   @SuppressWarnings({"rawtypes", "unchecked"})
   private static <C extends Comparable<C>> C rangeEndPointFromJson(Object o) {
     Exception e = null;
+    for (Class clsType : NUMERIC_VALUE_CLASSES) {
```

Comment:
doesn't the JSON include type information so you don't have to guess?
at least for literal, it seems to include it.

Reply:
The JSON includes the type information at the `RangeSet` level, but not down into the Range and rangeEndpoints levels. 
The rangeEndpoint only has `lower` and `upper`